source/common/router/config_impl.cc

```diff
@@ -1045,7 +1046,14 @@ RouteConstSharedPtr RegexRouteEntryImpl::matches(const Http::RequestHeaderMap& h
                                                  const StreamInfo::StreamInfo& stream_info,
                                                  uint64_t random_value) const {
   if (RouteEntryImplBase::matchRoute(headers, stream_info, random_value)) {
-    const absl::string_view path = Http::PathUtil::removeQueryAndFragment(headers.getPathValue());
+    absl::string_view path = Http::PathUtil::removeQueryAndFragment(headers.getPathValue());
+
+    std::string decoded_path;
+    if (percent_decode_input_) {
+      decoded_path = Http::Utility::PercentEncoding::decode(path);
```

Comment:
Should we instead be using `PathUtil::canonicalPath`? I.e. consistent percent decoding logic that is shared with the router lookup.

Reply:
Correct me if I'm wrong but I think `PathUtil::canonicalPath` does the inverse, ie. it applies the % encoding.

If I pass the path `[](){}! +*,:;<=>|@^'"&$\%` into `PathUtil::canonicalPath` what I get out is `[]()%7B%7D!%20+*,:;%3C=%3E%7C@%5E'%22&$%60/%`.

Reply:
Not quite, it rewrites the path using consistent rules after URL decoding it. It encodes characters that need to be encoded using uppercase hexadecmals, it writes characters that do not need to be % encoded as is. It handles correctly the case of double encoding, which this implementation does not. As such `/` will always be `%2F` and it is not needed to write disjunctions in regex matcher.
I think technically the problem you are presenting is already solved, i.e. you can write route matching using consistent rules without the need to accommodate differences in user-agents behavior. I think the documentation that explains these rules is missing, which is a problem.
The main problem with this implementation is that it does not handle double encoded sequences correctly, which is a security concern.